### PR TITLE
cassandra-stress: Silence NoSuchElementException warnings when iterator is exhausted

### PR DESCRIPTION
**Pull request by dkropachev**
Addressing having following errors when iterator exhausted:

java.util.NoSuchElementException
java.util.NoSuchElementException
java.util.NoSuchElementException
java.util.NoSuchElementException
java.util.NoSuchElementException
java.util.NoSuchElementException
java.io.IOException: Operation x10 on key(s) [49a26f35297303469236]: Error executing: (NoSuchElementException)
at org.apache.cassandra.stress.Operation.error(Operation.java:127)java.util.NoSuchElementException
at org.apache.cassandra.stress.Operation.timeWithRetry(Operation.java:105)com.datastax.driver.core.exceptions.TransportException: [/127.0.0.1:9042] Connection has been closed
java.util.NoSuchElementException
java.util.NoSuchElementException
java.util.NoSuchElementException
java.util.NoSuchElementException
java.util.NoSuchElementException

### CODE REVIEW

File: tools/stress/src/org/apache/cassandra/stress/Operation.java [hunk 1]
```diff
@@ -85,6 +86,10 @@ public final void timeWithRetry(RunOp run) throws IOException
                 success = run.run();
                 break;
             }
+            catch (NoSuchElementException e) {
+                // Pass thru iterator exhaustion exception
+                throw e;
+            }
```
Comment (penberg) [inline tools/stress/src/org/apache/cassandra/stress/Operation.java]: Why is this change needed?

File: tools/stress/src/org/apache/cassandra/stress/StressAction.java [hunk 2]
```diff
@@ -476,8 +477,10 @@ public void run()
                                 op.run(tclient);
                         }
                     }
-                    catch (Exception e)
-                    {
+                    catch (NoSuchElementException e) {
+                        // Silently reiterate when iterator is exhausted
```
Comment (penberg) [inline tools/stress/src/org/apache/cassandra/stress/StressAction.java]: Which iterator is exhausted? Why do we allow `next()` to be called if there's nothing to iterate?

File: tools/stress/src/org/apache/cassandra/stress/operations/userdefined/SchemaInsert.java [hunk 3]
```diff
@@ -69,30 +69,27 @@ public SchemaInsert(Timer timer, StressSettings settings, PartitionGenerator gen
...
+            List<BoundStatement> stmts;
 
-            if (stmts.size() == 0) {
-                for (PartitionIterator iterator : partitions)
-                    while (iterator.hasNext())
-                        stmts.add(bindRow(iterator.next()));
-            }
+            for (PartitionIterator iterator : partitions)
+                while (iterator.hasNext())
+                    stmts.add(bindRow(iterator.next()));
```
Comment (penberg) [inline tools/stress/src/org/apache/cassandra/stress/operations/userdefined/SchemaInsert.java]: Is this change related to the fix? Why is it needed?

File: tools/stress/src/org/apache/cassandra/stress/operations/userdefined/SchemaInsert.java [hunk 4]
```diff
@@ -69,30 +69,27 @@ public SchemaInsert(Timer timer, StressSettings settings, PartitionGenerator gen
...
 
             rowCount += stmts.size();
 
             // 65535 is max number of stmts per batch, so if we have more, we need to manually batch them
-            while (stmts.size() > 0)
+            for (int j = 0 ; j < stmts.size() ; j += 65535)
             {
-                List<BoundStatement> substmts = stmts.subList(0, Math.min(stmts.size(), 65535));
+                List<BoundStatement> substmts = stmts.subList(j, Math.min(j + stmts.size(), j + 65535));
```
Comment (penberg) [inline tools/stress/src/org/apache/cassandra/stress/operations/userdefined/SchemaInsert.java]: This looks like a performance optimization to avoid stmts.subList() at the bottom of this loop.

How is related to fixing the `NoSuchElement` spam in logs?

File: tools/stress/src/org/apache/cassandra/stress/operations/userdefined/SchemaQuery.java [hunk 5]
```diff
@@ -72,10 +71,7 @@ private JavaDriverRun(JavaDriverClient client)
 
         public boolean run() throws Exception
         {
-            if (bs == null) {
-                bs = bindArgs();
-            }
-            ResultSet rs = client.getSession().execute(bs);
+            ResultSet rs = client.getSession().execute(bindArgs());
```
Comment (penberg) [inline tools/stress/src/org/apache/cassandra/stress/operations/userdefined/SchemaQuery.java]: What is the purpose of this change?

**Comment by penberg** [comment]
@dkropachev Does this issue reproduce with Apache Cassandra `cassandra-stress`? If yes, let's open a Jira ticket and submit a patch to Cassandra too.

**Comment by dkropachev** [comment]
> @dkropachev Does this issue reproduce with Apache Cassandra `cassandra-stress`? If yes, let's open a Jira ticket and submit a patch to Cassandra too.
It is already there - https://github.com/apache/cassandra/pull/443
https://issues.apache.org/jira/browse/CASSANDRA-15034

This issue was initially addressed by Shlomi:
https://github.com/scylladb/scylla-dtest/issues/853
https://github.com/scylladb/scylla-tools-java/commit/e212e6269e3e86e63f2b814c52d74fabb6f3cd9a

But that fix did not cover all cases, but only ones that involved SchemaInsert and SchemaQuery.
This fix is general and fixes all possible cases, and also it rolls back changes from prior fix.

OperationIterator raises NoSuchElementException when it gets exhausted, as result it fails whole stress command.
Normally it never happens and should not happen. But when network issue pops up, logic does not reiterate cycle inside of StressAction$Consumer.run, but instead of it goes one more iteration inside of Operation.timeWithRetry, which opens up potential oportunity to call next() on empty iterator.

This fix passes exhausted iterator exception down to StressAction$Consumer.run cycle and silently reiterate there when it happens to replenish iterator with new values.

File: tools/stress/src/org/apache/cassandra/stress/operations/userdefined/SchemaQuery.java (hunk 5 above)
Comment (dkropachev) [reply tools/stress/src/org/apache/cassandra/stress/operations/userdefined/SchemaQuery.java]: It is part of rollback of e212e6269e3e86e63f2b814c52d74fabb6f3cd9a

File: tools/stress/src/org/apache/cassandra/stress/operations/userdefined/SchemaInsert.java (hunk 4 above)
Comment (dkropachev) [reply tools/stress/src/org/apache/cassandra/stress/operations/userdefined/SchemaInsert.java]: It is part of rollback of e212e6269e3e86e63f2b814c52d74fabb6f3cd9a

File: tools/stress/src/org/apache/cassandra/stress/operations/userdefined/SchemaInsert.java (hunk 3 above)
Comment (dkropachev) [reply tools/stress/src/org/apache/cassandra/stress/operations/userdefined/SchemaInsert.java]: It is part of rollback of e212e6269e3e86e63f2b814c52d74fabb6f3cd9a

File: tools/stress/src/org/apache/cassandra/stress/StressAction.java (hunk 2 above)
Comment (dkropachev) [reply tools/stress/src/org/apache/cassandra/stress/StressAction.java]: https://github.com/scylladb/scylla-tools-java/pull/143#issuecomment-584769048

File: tools/stress/src/org/apache/cassandra/stress/Operation.java (hunk 1 above)
Comment (dkropachev) [reply tools/stress/src/org/apache/cassandra/stress/Operation.java]: https://github.com/scylladb/scylla-tools-java/pull/143#issuecomment-584769048

Comment (dkropachev) [reply tools/stress/src/org/apache/cassandra/stress/Operation.java]: To pass this exception down to StressAction$Consumer.run cycle

File: tools/stress/src/org/apache/cassandra/stress/StressAction.java (hunk 2 above)
Comment (dkropachev) [reply tools/stress/src/org/apache/cassandra/stress/StressAction.java]: PartitionOperation.partitions or partitionIterator enclosed in it.

**Comment by dkropachev** [comment]
Hello Guys, Can I have it reviewed ?

**Comment by penberg** [comment]
Please update the commit message to include the information that's now in the comments.

The commit is still hard to understand, perhaps rebasing it on top of a `git revert e212e6269e3e86e63f2b814c52d74fabb6f3cd9a` makes it more readable.

**Comment by dkropachev** [comment]
@penberg , Great idea. take a look now.

**Comment by penberg** [comment]
Merged the fix as fe4391ba8a573a6376a4fa3cc6a39dea1291ecfb.

**Comment by penberg** [comment]
@dkropachev Btw, reverts should have some explanation _why_ a commit is being reverted. I added an explanation in the revert I did in https://github.com/scylladb/scylla-tools-java/commit/a877883ad896b2858c7447c49d60974bafe4b8c1. I hope my explanation is correct.
